### PR TITLE
Implement Three New Plugins

### PR DESCRIPTION
**Pull request by SerperiorBae**
- Create Sapience-Plugins.js
- Implement Animals Plugin
- Implement Giphy Command
- Implement Events Plugin

### CODE REVIEW

File: chat-plugins/sapience-plugins.js [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+/****************************************
+* Events Plugin by RoboPhill & Sapience *
+****************************************/
+
+    event: {
+        create: function (target, room, user) {
+            let params = target.split(", ") || target.split(",");
+            if (!target || !params[0] || !params[1] || !params[2]) return this.errorReply('Usage: /event create [Name], [Desc], [Room]');
+            if (params[1].length > 500) return this.errorReply('Descriptions must be a maximum of 500 characters!');
+            if (EventsOn[toId(params[0])]) return this.sendReply("/event - The event \"" + params[0] + "\" already exists.");
```
Comment (SilverTactic) [inline chat-plugins/sapience-plugins.js]: Even though you added this check, you didn't set the value of the <code>toId(params[0])</code> key anywhere else in the script. You only set values for <code>Name</code>, <code>DisplayName</code>, <code>Desc</code>, and <code>Room</code>

File: chat-plugins/sapience-plugins.js [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+	let eventList = JSON.parse(fs.readFileSync('config/events.json'));
+	var key, count = 0;
+	for(key in eventList) {
+		if(eventList.hasOwnProperty(key)) {
+			count++;
+		}
+	}
+	eventList[count] = EventsOn;
+	fs.writeFileSync('config/events.json', JSON.stringify(eventList));
+	Object.assign(CommandParser.commands, EventsOn);
```
Comment (SilverTactic) [inline chat-plugins/sapience-plugins.js]: Commands are supposed to be functions, but your <code>EventsOn</code> object keys aren't functions :/

File: chat-plugins/sapience-plugins.js [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+	} catch (e) {}
+}
+
+setTimeout(function load() {
+	loadEvents();
+}, 1000);
+
+function saveEvents() {
+	let eventList = JSON.parse(fs.readFileSync('config/events.json'));
+	var key, count = 0;
```
Comment (panpawn) [inline chat-plugins/sapience-plugins.js]: `let`

File: chat-plugins/sapience-plugins.js [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+	for(key in eventList) {
```
Comment (panpawn) [inline chat-plugins/sapience-plugins.js]: `for (key...`

File: chat-plugins/sapience-plugins.js [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+		if(eventList.hasOwnProperty(key)) {
```
Comment (panpawn) [inline chat-plugins/sapience-plugins.js]: `if (...`

File: chat-plugins/sapience-plugins.js [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+	}
+	eventList[count] = EventsOn;
+	fs.writeFileSync('config/events.json', JSON.stringify(eventList));
+}
+
+exports.commands = {
+
+// Animal command by Sapience
+	animal: 'animals',
+	animals: function(target, room, user) {
```
Comment (panpawn) [inline chat-plugins/sapience-plugins.js]: `function (...`

File: chat-plugins/sapience-plugins.js [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+	        }
+			
+	        this.sendReply('|raw|' + display)
+    },
+    },
+    eventhelp: ['|raw|<div class="infobox">Event Command by <font style="color: orange">Sapience</font> & <font style="color: rgb(170, 100, 8)">RoboPhill</font><br>' +
+	            '<ul><li> /event create [name], [description], [room] - Creates, Stores and Declares the event <small>Don\'t use commas in Desc</small></li>' + 
+	            '<li> /event remove [name] - Removes said event from being displayed</li>' + 
+	            '<li> /event display - Displays all events on at this time</li></div>'],
+};
```
Comment (panpawn) [inline chat-plugins/sapience-plugins.js:226]: add trailing whitespace here

File: chat-plugins/sapience-plugins.js [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+	        for (let event in eventList) {
+				
+					display += '<tr><td style="border: solid, 1px, #ffc775;">' + eventList[event]["DisplayName"] + '</td><td style="border: solid, 1px, #ffc775;">' + eventList[event]["Desc"] + '</td><td style="border: solid, 1px, #ffc775;"><button style="background-color: #ff560e ; border: solid, 1px, #ffc775 ; border-radius: 15px 0px ; color: #ffc775" name="joinRoom" value="' + toId(eventList[event]["Room"]) + '">'  + eventList[event]["Room"] + '</button></td></tr><tr>';
+	        }
+			
+	        this.sendReply('|raw|' + display)
+    },
+    },
+    eventhelp: ['|raw|<div class="infobox">Event Command by <font style="color: orange">Sapience</font> & <font style="color: rgb(170, 100, 8)">RoboPhill</font><br>' +
```
Comment (panpawn) [inline chat-plugins/sapience-plugins.js]: don't do this like this

File: chat-plugins/sapience-plugins.js [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+	        /*if (!EventsOn[params]) return this.sendReply("/event - The Event: \"" + params + "\" does not exist, or was added manually.");
+	        	this.sendReply("The Event: \"" + params + "\" has been removed.");
+				this.logModCommand(user.name + " removed the Event " + params);
+				delete CommandParser.commands[params];
+				delete EventsOn[params];
+				saveEvents();*/
+	    },
+	    display: function (target, room, user) {
+	        if (!this.runBroadcast()) return;
+			if (!fs.existsSync('config/events.json')){
```
Comment (panpawn) [inline chat-plugins/sapience-plugins.js]: spacing

File: chat-plugins/sapience-plugins.js [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+				return this.sendReply("No events could be found");
+			}
+			let eventList = JSON.parse(fs.readFileSync('config/events.json'))
+			if (Object.keys(EventsOn).length < 1);
```
Comment (panpawn) [inline chat-plugins/sapience-plugins.js:206]: ????

File: chat-plugins/sapience-plugins.js [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+			if(!eventList) return this.sendReply("No events could be found");
+	        let display = '<center><strong style="font-size: 20px; font-weight: bold;">Events Running</strong><br><table style="border: solid, 1px, #ffc775; border-collapse: collapse; text-align: center; width:100%;"><tr><th style="background-color: #ff560e ; border: solid, 1px, #ffc775; width: 20%;">Event Name</th><th style="background-color: #ff560e ; border: solid, 1px, #ffc775; width: 60%;">Description</th><th style="background-color: #ff560e ; border: solid, 1px, #ffc775; width: 20%;">Event Room</th>';
+	        for (let event in eventList) {
+
```
Comment (panpawn) [inline chat-plugins/sapience-plugins.js]: spacing

File: chat-plugins/sapience-plugins.js [hunk 12]
```diff
@@ -0,0 +1,198 @@
...
+			if (Object.keys(EventsOn).length < 1);
+			if(!eventList) return this.sendReply("No events could be found");
+	        let display = '<center><strong style="font-size: 20px; font-weight: bold;">Events Running</strong><br><table style="border: solid, 1px, #ffc775; border-collapse: collapse; text-align: center; width:100%;"><tr><th style="background-color: #ff560e ; border: solid, 1px, #ffc775; width: 20%;">Event Name</th><th style="background-color: #ff560e ; border: solid, 1px, #ffc775; width: 60%;">Description</th><th style="background-color: #ff560e ; border: solid, 1px, #ffc775; width: 20%;">Event Room</th>';
+	        for (let event in eventList) {
+				
+					display += '<tr><td style="border: solid, 1px, #ffc775;">' + eventList[event]["DisplayName"] + '</td><td style="border: solid, 1px, #ffc775;">' + eventList[event]["Desc"] + '</td><td style="border: solid, 1px, #ffc775;"><button style="background-color: #ff560e ; border: solid, 1px, #ffc775 ; border-radius: 15px 0px ; color: #ffc775" name="joinRoom" value="' + toId(eventList[event]["Room"]) + '">'  + eventList[event]["Room"] + '</button></td></tr><tr>';
```
Comment (panpawn) [inline chat-plugins/sapience-plugins.js]: spacing

File: chat-plugins/sapience-plugins.js [hunk 13]
```diff
@@ -0,0 +1,198 @@
...
+	        }
+
```
Comment (panpawn) [inline chat-plugins/sapience-plugins.js]: remove this line

File: chat-plugins/sapience-plugins.js [hunk 14]
```diff
@@ -0,0 +1,198 @@
...
+			if(!eventList) return this.sendReply("No events could be found");
+	        let display = '<center><strong style="font-size: 20px; font-weight: bold;">Events Running</strong><br><table style="border: solid, 1px, #ffc775; border-collapse: collapse; text-align: center; width:100%;"><tr><th style="background-color: #ff560e ; border: solid, 1px, #ffc775; width: 20%;">Event Name</th><th style="background-color: #ff560e ; border: solid, 1px, #ffc775; width: 60%;">Description</th><th style="background-color: #ff560e ; border: solid, 1px, #ffc775; width: 20%;">Event Room</th>';
+	        for (let event in eventList) {
+				
+					display += '<tr><td style="border: solid, 1px, #ffc775;">' + eventList[event]["DisplayName"] + '</td><td style="border: solid, 1px, #ffc775;">' + eventList[event]["Desc"] + '</td><td style="border: solid, 1px, #ffc775;"><button style="background-color: #ff560e ; border: solid, 1px, #ffc775 ; border-radius: 15px 0px ; color: #ffc775" name="joinRoom" value="' + toId(eventList[event]["Room"]) + '">'  + eventList[event]["Room"] + '</button></td></tr><tr>';
+	        }
+			
+	        this.sendReply('|raw|' + display)
+    },
```
Comment (panpawn) [inline chat-plugins/sapience-plugins.js]: spacing

File: chat-plugins/sapience-plugins.js [hunk 15]
```diff
@@ -0,0 +1,198 @@
...
+			if (Object.keys(EventsOn).length < 1);
+			if(!eventList) return this.sendReply("No events could be found");
+	        let display = '<center><strong style="font-size: 20px; font-weight: bold;">Events Running</strong><br><table style="border: solid, 1px, #ffc775; border-collapse: collapse; text-align: center; width:100%;"><tr><th style="background-color: #ff560e ; border: solid, 1px, #ffc775; width: 20%;">Event Name</th><th style="background-color: #ff560e ; border: solid, 1px, #ffc775; width: 60%;">Description</th><th style="background-color: #ff560e ; border: solid, 1px, #ffc775; width: 20%;">Event Room</th>';
+	        for (let event in eventList) {
+				
+					display += '<tr><td style="border: solid, 1px, #ffc775;">' + eventList[event]["DisplayName"] + '</td><td style="border: solid, 1px, #ffc775;">' + eventList[event]["Desc"] + '</td><td style="border: solid, 1px, #ffc775;"><button style="background-color: #ff560e ; border: solid, 1px, #ffc775 ; border-radius: 15px 0px ; color: #ffc775" name="joinRoom" value="' + toId(eventList[event]["Room"]) + '">'  + eventList[event]["Room"] + '</button></td></tr><tr>';
+	        }
+			
+	        this.sendReply('|raw|' + display)
```
Comment (panpawn) [inline chat-plugins/sapience-plugins.js]: add semi colon

File: chat-plugins/sapience-plugins.js [hunk 16]
```diff
@@ -0,0 +1,198 @@
...
+	animalshelp: ['|html|<div class="infobox">Animals Plugin by <font color="orange">Sapience</font>' +
+		'<ul><li>/animals cat - Displays a cat.</li>' +
+		'<li>/animals kitten - Displays a kitten.</li>' +
+		'<li>/animals dog - Displays a dog.</li>' +
+		'<li>/animals puppy - Displays a puppy.</li>' +
+		'<li>/animals bunny - Displays a bunny.</li>' +
+		'<li>/animals otter - Displays an otter.</li>' +
+		'<li>/animals pokemon - Displays a pokemon.</li>' +
+		'<li>/animals help - Displays this help box.</li></ul></div>'
+],
```
Comment (panpawn) [inline chat-plugins/sapience-plugins.js]: put this on the line above

File: chat-plugins/sapience-plugins.js [hunk 17]
```diff
@@ -0,0 +1,198 @@
...
+	gif: 'giphy',
+	giphy: function(target, room, user) {
```
Comment (panpawn) [inline chat-plugins/sapience-plugins.js]: `function (...`

File: chat-plugins/sapience-plugins.js [hunk 18]
```diff
@@ -0,0 +1,198 @@
...
+
+    event: {
+        create: function (target, room, user) {
+            let params = target.split(", ") || target.split(",");
+            if (!target || !params[0] || !params[1] || !params[2]) return this.errorReply('Usage: /event create [Name], [Desc], [Room]');
+            if (params[1].length > 500) return this.errorReply('Descriptions must be a maximum of 500 characters!');
+            if (EventsOn[toId(params[0])]) return this.sendReply("/event - The event \"" + params[0] + "\" already exists.");
+            if (!this.can('lock')) return false;
+            if (!target) return;
+			if (!fs.existsSync('config/events.json')){
```
Comment (panpawn) [inline chat-plugins/sapience-plugins.js]: spacing

Comment (panpawn) [inline chat-plugins/sapience-plugins.js]: `if (...) {`

File: chat-plugins/sapience-plugins.js [hunk 19]
```diff
@@ -0,0 +1,198 @@
...
+			}
+			let eventList = JSON.parse(fs.readFileSync('config/events.json'));
+			for (let event in eventList){
+				if (eventList[event]["Name"] == params[0]){
+					return this.sendReply("There is already an Event named " + params[0]);
+					break;
+				}
+			}
+            for (let id in Rooms.rooms) {
+				if (id !== 'global') Rooms.rooms[id].addRaw('<div style="background-color: #ff560e ; border: solid, 1px, #ffc775 ; border-top-left-radius: 0em ; border-top-right-radius: 6em ; color: #ffc775 ; font-size: 20px; text-align: center ; text-decoration: underline ; padding: 4px ;">' +
```
Comment (panpawn) [inline chat-plugins/sapience-plugins.js]: spacing

File: chat-plugins/sapience-plugins.js [hunk 20]
```diff
@@ -0,0 +1,198 @@
...
+			let eventList = JSON.parse(fs.readFileSync('config/events.json'));
+			for (let event in eventList){
+				if (eventList[event]["Name"] == params[0]){
+					return this.sendReply("There is already an Event named " + params[0]);
+					break;
+				}
+			}
+            for (let id in Rooms.rooms) {
+				if (id !== 'global') Rooms.rooms[id].addRaw('<div style="background-color: #ff560e ; border: solid, 1px, #ffc775 ; border-top-left-radius: 0em ; border-top-right-radius: 6em ; color: #ffc775 ; font-size: 20px; text-align: center ; text-decoration: underline ; padding: 4px ;">' + 
+                                                            params[0] + '</div><div style="background-color: #ff560e ; border-left: solid, 1px, #ffc775 ; border-right: solid, 1px, #ffc775 ; font-size: 12px ; text-align: center ; padding: 8px, 4px ;">' +
```
Comment (panpawn) [inline chat-plugins/sapience-plugins.js]: these r spaces and not tabs..

File: chat-plugins/sapience-plugins.js [hunk 21]
```diff
@@ -0,0 +1,198 @@
...
+                                                            params[1] + '</div><div style="background-color: #ff560e ; border: solid, 1px, #ffc775 ; border-bottom-left-radius: 6em ; border-bottom-right-radius: 0em ; color: #ffc775 ; font-size: 20px; text-align: center ; text-decoration: underline ; padding: 4px ;"><button style="background-color: #ffc775 ; border: solid, 1px, #ff560e ; border-radius: 15px 0px ; color: #ff560e" name="joinRoom" value="' +
+                                                            toId(params[2]) + '">' + params[0] + ' Event in here!</button></div>');
+            }
+            this.logModCommand(user.name + " created event in: " + params[2]);
+            		EventsOn["Name"] = toId(params[0]);
+			EventsOn["DisplayName"] = params[0];
+			EventsOn["Desc"] = params[1];
+			EventsOn["Room"] = params[2];
+			saveEvents();
+
```
Comment (panpawn) [inline chat-plugins/sapience-plugins.js]: remove this line

File: chat-plugins/sapience-plugins.js [hunk 22]
```diff
@@ -0,0 +1,198 @@
...
+				
+	    },
+	    remove: function (target, room, user) {
+	        if (!this.can('lock')) return false;
+	        let params = toId(target);
+	        if (!target) return this.errorReply('Usage: /event remove [Name]');
+			let eventList = JSON.parse(fs.readFileSync('config/events.json'));
+			console.log(params);
+			if (JSON.stringify(eventList).indexOf(params) >= 0) {
+				for (let event in eventList){
```
Comment (panpawn) [inline chat-plugins/sapience-plugins.js]: `for (...) {`

File: chat-plugins/sapience-plugins.js [hunk 23]
```diff
@@ -0,0 +1,198 @@
...
+					if (eventList[event]["Name"] == params || eventList[event]["DisplayName"] == params){
```
Comment (panpawn) [inline chat-plugins/sapience-plugins.js]: `if (...) {`

File: chat-plugins/sapience-plugins.js [hunk 24]
```diff
@@ -0,0 +1,198 @@
...
+						this.sendReply("The Event: \"" + params + "\" has been removed.");
+						this.logModCommand(user.name + " removed the Event " + params);
+						delete CommandParser.commands[params];
+						delete eventList[event];
+						fs.writeFileSync('config/events.json', JSON.stringify(eventList));
+						break;
+					}
+				}
+			}else{
```
Comment (panpawn) [inline chat-plugins/sapience-plugins.js]: `} else {`

File: chat-plugins/sapience-plugins.js [hunk 25]
```diff
@@ -0,0 +1,198 @@
...
+				return this.sendReply("/event - The Event: \"" + params + "\" does not exist, or was added manually.");
+			}
+	        /*if (!EventsOn[params]) return this.sendReply("/event - The Event: \"" + params + "\" does not exist, or was added manually.");
```
Comment (panpawn) [inline chat-plugins/sapience-plugins.js]: should this still be here?

File: chat-plugins/sapience-plugins.js [hunk 26]
```diff
@@ -0,0 +1,198 @@
...
+	        	this.sendReply("The Event: \"" + params + "\" has been removed.");
+				this.logModCommand(user.name + " removed the Event " + params);
+				delete CommandParser.commands[params];
+				delete EventsOn[params];
+				saveEvents();*/
+	    },
+	    display: function (target, room, user) {
+	        if (!this.runBroadcast()) return;
```
Comment (panpawn) [inline chat-plugins/sapience-plugins.js]: you and your spaces

**Comment by panpawn** [comment]
There's a lot of spacing issues with this.

**Comment by SilverTactic** [comment]
<a href = "http://jsbeautifier.org">jsbeautifier.org</a> could help you a bit I guess

**Comment by panpawn** [comment]
can you squash?

**Comment by panpawn** [comment]
I just read this again. 

There is no way this is getting merged,  mostly because of this "giphy" command.  The fact of the matter is this command is a hige liability because you have no good way of filtering it. 

Sorry,  I should of caught on to that sooner. . Don't bother continuing to work on this pr.

**Pull request closed without merging**